### PR TITLE
[MeshingApplication] AtomicUtilities/ParallelUtilities migration

### PR DESCRIPTION
**Pull request by loumalouomega**
**Description**
Depends on #8350. This PR adds the AtomicUtilities/ParallelUtilities to all my implementations in the MeshingApplication. NOTE: We need to create "CriticalUtilities"

**Changelog**
- Using AtomicUtilities
- Using ParallelUtilities

### CODE REVIEW

**Review by philbucher — COMMENTED** [review]
I have some important comments, but not much time first half of the week, please wait for my review

**Comment by loumalouomega** [comment]
> I have some important comments, but not much time first half of the week, please wait for my review

Of course, no hurry, I used this app to practice with the parallel utilities

File: applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp [hunk 1]
```diff
@@ -3886,16 +3894,20 @@ void MmgUtilities<TMMGLibrary>::GenerateMeshDataFromModelPart(
...
     }
-    counter_to_remesh = 0;
-    #pragma omp parallel for reduction(+: counter_to_remesh)
-    for(int i = 0; i < static_cast<int>(r_elements_array.size()); ++i) {
-        auto it_elem = it_elem_begin + i;
 
-        const bool old_entity = it_elem->IsDefined(OLD_ENTITY) ? it_elem->Is(OLD_ENTITY) : false;
+    /* Elements */
+    counter_to_remesh = 0;
+    counter_to_remesh = block_for_each<SumReduction<IndexType>>(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp]:
```suggestion
    counter_to_remesh = block_for_each<SumReduction<IndexType>>(
```

File: applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp [hunk 2]
```diff
@@ -3861,16 +3865,20 @@ void MmgUtilities<TMMGLibrary>::GenerateMeshDataFromModelPart(
...
     }
-    counter_to_remesh = 0;
-    #pragma omp parallel for reduction(+: counter_to_remesh)
-    for(int i = 0; i < static_cast<int>(r_conditions_array.size()); ++i) {
-        auto it_cond = it_cond_begin + i;
 
-        const bool old_entity = it_cond->IsDefined(OLD_ENTITY) ? it_cond->Is(OLD_ENTITY) : false;
+    /* Conditions */
+    counter_to_remesh = 0;
+    counter_to_remesh = block_for_each<SumReduction<IndexType>>(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp]:
```suggestion
    counter_to_remesh = block_for_each<SumReduction<IndexType>>(
```

File: applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp [hunk 3]
```diff
@@ -3824,28 +3825,31 @@ void MmgUtilities<TMMGLibrary>::GenerateMeshDataFromModelPart(
...
 
     // We reorder the ids to avoid conflicts with the rest (using as reference the OLD_ENTITY)
+    /* Nodes */
     IndexType counter_to_remesh = 0;
-    #pragma omp parallel for reduction(+: counter_to_remesh)
-    for(int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
-        auto it_node = it_node_begin + i;
-
-        const bool old_entity = it_node->IsDefined(OLD_ENTITY) ? it_node->Is(OLD_ENTITY) : false;
+    counter_to_remesh = block_for_each<SumReduction<IndexType>>(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp]:
```suggestion
    IndexType counter_to_remesh = block_for_each<SumReduction<IndexType>>(
```

File: applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp [hunk 4]
```diff
@@ -3824,28 +3825,31 @@ void MmgUtilities<TMMGLibrary>::GenerateMeshDataFromModelPart(
     }
 
     // Set flag on nodes
-    #pragma omp parallel for
-    for (int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+    IndexPartition<std::size_t>(r_nodes_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_utilities/mmg/mmg_utilities.cpp]: `block_for_each`

File: applications/MeshingApplication/custom_utilities/meshing_utilities.cpp [hunk 5]
```diff
@@ -72,13 +72,12 @@ void ComputeElementsSize(ModelPart& rModelPart)
     // Iterating over the elements
     ElementsArrayType& r_elements_array = rModelPart.Elements();
     const auto it_elem_begin= r_elements_array.begin();
-    const int number_of_elements = static_cast<int>(r_elements_array.size());
 
-    #pragma omp parallel for
-    for (int i = 0; i < number_of_elements; ++i) {
+    IndexPartition<std::size_t>(r_elements_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_utilities/meshing_utilities.cpp]: block_for_Each

File: applications/MeshingApplication/custom_utilities/meshing_utilities.cpp [hunk 6]
```diff
@@ -46,10 +47,9 @@ void BlockThresholdSizeElements(
...
```
Comment (philbucher) [inline applications/MeshingApplication/custom_utilities/meshing_utilities.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp [hunk 7]
```diff
@@ -362,26 +347,25 @@ void NodalValuesInterpolationProcess<TDim>::ComputeNormalSkin(ModelPart& rModelP
...
             }
         }
-    }
+    });
 
+    // Iterate over nodes
     NodesArrayType& r_nodes_array = rModelPart.Nodes();
-    const int num_nodes = static_cast<int>(r_nodes_array.size());
     const auto it_node_begin = r_nodes_array.begin();
+    IndexPartition<std::size_t>(r_nodes_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp [hunk 8]
```diff
@@ -362,26 +347,25 @@ void NodalValuesInterpolationProcess<TDim>::ComputeNormalSkin(ModelPart& rModelP
             const array_1d<double, 3> normal = this_geometry.UnitNormal(aux_coords);
             auto& aux_normal = this_node.GetValue(NORMAL);
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp:346]: better not use auto but specify explicitly the type

File: applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp [hunk 9]
```diff
@@ -362,26 +347,25 @@ void NodalValuesInterpolationProcess<TDim>::ComputeNormalSkin(ModelPart& rModelP
...
             for (IndexType index = 0; index < 3; ++index) {
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp:347]: then you don't need this loop but can call `AtomicAdd` directly

File: applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp [hunk 10]
```diff
@@ -343,9 +328,9 @@ void NodalValuesInterpolationProcess<TDim>::ComputeNormalSkin(ModelPart& rModelP
     ConditionsArrayType& r_conditions_array = rModelPart.Conditions();
     const auto it_cond_begin = r_conditions_array.begin();
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_conditions_array.size()); ++i) {
-        auto it_cond = it_cond_begin + i;
+    IndexPartition<std::size_t>(r_conditions_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp [hunk 11]
```diff
@@ -283,9 +266,11 @@ void NodalValuesInterpolationProcess<TDim>::ExtrapolateValues(
         }
     }
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(point_list_destination.size()); ++i)
-        point_list_destination[i]->UpdatePoint();
+    // Update point
+    IndexPartition<std::size_t>(point_list_destination.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/nodal_values_interpolation_process.cpp:270]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 12]
```diff
@@ -1191,16 +1196,16 @@ void MmgProcess<TMMGLibrary>::CleanSuperfluousNodes()
     const auto it_elem_begin = r_elements_array.begin();
 
     // Saving the nodes that belong to an element
-    #pragma omp parallel for
-    for(int i_elem = 0; i_elem < static_cast<int>(r_elements_array.size()); ++i_elem ){
+    IndexPartition<std::size_t>(r_elements_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 13]
```diff
@@ -1117,11 +1122,11 @@ void MmgProcess<TMMGLibrary>::CreateDebugPrePostRemeshOutput(ModelPart& rOldMode
     auto& r_elements_array_2 = r_copy_old_model_part.Elements();
     const auto it_elem_begin_2 = r_elements_array_2.begin();
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_elements_array_2.size()); ++i) {
+    IndexPartition<std::size_t>(r_elements_array_2.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 14]
```diff
@@ -1103,11 +1108,11 @@ void MmgProcess<TMMGLibrary>::CreateDebugPrePostRemeshOutput(ModelPart& rOldMode
     auto& r_elements_array_1 = r_auxiliar_model_part.Elements();
     const auto it_elem_begin_1 = r_elements_array_1.begin();
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_elements_array_1.size()); ++i) {
+    IndexPartition<std::size_t>(r_elements_array_1.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 15]
```diff
@@ -942,18 +945,19 @@ void MmgProcess<TMMGLibrary>::ExtrudeTrianglestoPrisms(ModelPart& rOldModelPart)
         }
     }
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(num_nodes); ++i) {
+    IndexPartition<std::size_t>(r_nodes_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 16]
```diff
@@ -906,30 +911,28 @@ void MmgProcess<TMMGLibrary>::ExtrudeTrianglestoPrisms(ModelPart& rOldModelPart)
...
 
     // Reset NORMAL
     VariableUtils().SetNonHistoricalVariableToZero(NORMAL, r_nodes_array);
 
     // Declare auxiliar coordinates
     GeometryType::CoordinatesArrayType aux_coords;
 
-    #pragma omp parallel for firstprivate(aux_coords)
-    for(int i = 0; i < static_cast<int>(num_elements); ++i) {
+    IndexPartition<std::size_t>(r_elements_array.size()).for_each(aux_coords,
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 17]
```diff
@@ -712,16 +715,18 @@ void MmgProcess<TMMGLibrary>::InitializeElementsAndConditions()
...
 
     // Iterate over elements
     auto& r_elements_array = mrThisModelPart.Elements();
     const auto it_elem_begin = r_elements_array.begin();
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_elements_array.size()); ++i)
+    IndexPartition<std::size_t>(r_elements_array.size()).for_each(
+        [&it_elem_begin,&r_process_info](std::size_t i) {
         (it_elem_begin + i)->Initialize(r_process_info);
+    });
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: EntitiesUtilties

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 18]
```diff
@@ -659,14 +662,14 @@ void MmgProcess<TMMGLibrary>::ExecuteRemeshing()
 
             const array_1d<double, 3> zero_vector = ZeroVector(3);
 
-            #pragma omp parallel for
-            for(int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+            IndexPartition<std::size_t>(r_nodes_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 19]
```diff
@@ -644,13 +647,13 @@ void MmgProcess<TMMGLibrary>::ExecuteRemeshing()
             r_nodes_array = mrThisModelPart.Nodes();
             const auto it_node_begin = r_nodes_array.begin();
 
-            #pragma omp parallel for
-            for(int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+            IndexPartition<std::size_t>(r_nodes_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 20]
```diff
@@ -601,12 +603,13 @@ void MmgProcess<TMMGLibrary>::ExecuteRemeshing()
     /* Unmoving the original mesh to be able to interpolate */
     if (mFramework == FrameworkEulerLagrange::LAGRANGIAN) {
         NodesArrayType& r_old_nodes_array = r_old_model_part.Nodes();
+        const auto it_old_node_begin = r_old_nodes_array.begin();
 
-        #pragma omp parallel for
-        for(int i = 0; i < static_cast<int>(r_old_nodes_array.size()); ++i) {
-            auto it_node = r_old_nodes_array.begin() + i;
+        IndexPartition<std::size_t>(r_old_nodes_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 21]
```diff
@@ -601,12 +603,13 @@ void MmgProcess<TMMGLibrary>::ExecuteRemeshing()
...
+        [&it_old_node_begin](std::size_t i) {
+            auto it_node = it_old_node_begin + i;
             noalias(it_node->Coordinates()) = it_node->GetInitialPosition().Coordinates();
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: doesn't this exist in the variable utils?

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 22]
```diff
@@ -512,45 +514,45 @@ void MmgProcess<TMMGLibrary>::ExecuteRemeshing()
...
+    });
     r_old_model_part.AddConditions( mrThisModelPart.ConditionsBegin(), mrThisModelPart.ConditionsEnd() );
     mrThisModelPart.RemoveConditionsFromAllLevels(TO_ERASE);
 
     auto& r_elements_array = mrThisModelPart.Elements();
     const auto it_elem_begin = r_elements_array.begin();
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_elements_array.size()); ++i) {
+    IndexPartition<std::size_t>(r_elements_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp [hunk 23]
```diff
@@ -512,45 +514,45 @@ void MmgProcess<TMMGLibrary>::ExecuteRemeshing()
...
+    });
     r_old_model_part.AddNodes( mrThisModelPart.NodesBegin(), mrThisModelPart.NodesEnd() );
     mrThisModelPart.RemoveNodesFromAllLevels(TO_ERASE);
 
     auto& r_conditions_array = mrThisModelPart.Conditions();
     const auto it_cond_begin = r_conditions_array.begin();
 
-    #pragma omp parallel for
-    for(int i = 0; i < static_cast<int>(r_conditions_array.size()); ++i) {
+    IndexPartition<std::size_t>(r_conditions_array.size()).for_each(
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/mmg/mmg_process.cpp]: block_for_each

File: applications/MeshingApplication/custom_processes/metrics_levelset_process.cpp [hunk 24]
```diff
@@ -80,37 +79,42 @@ void ComputeLevelSetSolMetricProcess<TDim>::Execute()
...
-        #pragma omp parallel for
-        for(int i = 0; i < num_nodes; ++i) {
-            auto it_node = it_node_begin + i;
-            it_node->SetValue(r_tensor_variable, aux_zero_vector);
-        }
+        IndexPartition<std::size_t>(r_nodes_array.size()).for_each(
+        [&it_node_begin,&r_tensor_variable](std::size_t i_node) {
+            auto it_node = it_node_begin + i_node;
+            it_node->SetValue(r_tensor_variable, r_tensor_variable.Zero());
+        });
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/metrics_levelset_process.cpp]: variableUtils

**Review by philbucher — COMMENTED** [review]
can you please use `block_for_each` instead of `IndexPartition`?
We generally should only use the `IndexPartition` if we cannot use the block version

**Comment by loumalouomega** [comment]
> can you please use `block_for_each` instead of `IndexPartition`?
> We generally should only use the `IndexPartition` if we cannot use the block version

Ok, I used IndexPartition because it was easier to migrate as I use it_whatever a lot

**Comment by philbucher** [comment]
> 
> 
> > can you please use `block_for_each` instead of `IndexPartition`?
> > We generally should only use the `IndexPartition` if we cannot use the block version
> 
> Ok, I used IndexPartition because it was easier to migrate as I use it_whatever a lot

Yes I know, but it is easier to use and now I think is a good opportunity to clean up

**Comment by loumalouomega** [comment]
@philbucher I think this is ready for final review

File: applications/MeshingApplication/custom_processes/metrics_error_process.cpp [hunk 25]
```diff
@@ -138,22 +141,24 @@ void MetricErrorProcess<TDim>::CalculateMetric()
     }
 
     // Iteration over all nodes
-    const int num_nodes = static_cast<int>(r_nodes_array.size());
-    KRATOS_DEBUG_ERROR_IF(num_nodes == 0) <<  "ERROR:: Empty list of nodes" << std::endl;
+    KRATOS_DEBUG_ERROR_IF(r_nodes_array.size() == 0) <<  "ERROR:: Empty list of nodes" << std::endl;
+
+    // Auxiliar variables
+    const auto average_nodal_h = mAverageNodalH;
+    const auto echo_level = mEchoLevel;
```
Comment (philbucher) [inline applications/MeshingApplication/custom_processes/metrics_error_process.cpp:143]: why is this necessary?
Cannot you capture those too?

Comment (philbucher) [reply applications/MeshingApplication/custom_processes/metrics_error_process.cpp:143]: or use `this`?

Comment (loumalouomega) [reply applications/MeshingApplication/custom_processes/metrics_error_process.cpp:143]: ...that's genious...

**Review by philbucher — APPROVED** [review]
This is a large PR so I hope I didn't miss anything
Looks good to me!
Thx

**Comment by marcnunezc** [comment]
@loumalouomega , I compiled master as I wanted to update pmmg files as well and the cpp tests in the meshing App are giving me a seg fault (it seems it is coming from the interpolation process). Are they working for you?

@philbucher, are the cpp tests included in the CI?

**Comment by loumalouomega** [comment]
I run the tests...in Release...

**Comment by loumalouomega** [comment]
> @loumalouomega , I compiled master as I wanted to update pmmg files as well and the cpp tests in the meshing App are giving me a seg fault (it seems it is coming from the interpolation process). Are they working for you?
> 
> @philbucher, are the cpp tests included in the CI?

Can you take a look?

**Comment by loumalouomega** [comment]
> @philbucher, are the cpp tests included in the CI?

yes

**Comment by philbucher** [comment]
> > @philbucher, are the cpp tests included in the CI?
> 
> yes

yes, in release & fulldebug

but with only 2 cores

**Comment by marcnunezc** [comment]
Ok, I might be doing something wrong then. 

I can say that this is happening if I compile with the StrucutralApp, as the segfault comes from this test:

**Comment by marcnunezc** [comment]
The problem was I was not compiling the CLawApp, sorry for the noise, see #8374.

**Comment by philbucher** [comment]
We should add a check for that and skip the test in such a case
@marcnunezc since you know how to reproduce could you add it?
Thx

**Comment by marcnunezc** [comment]
Yes yes, I am adding int #8374

**Comment by philbucher** [comment]
I just saw, thx
